### PR TITLE
modules: openthread: add `otPlatDiagRadioTransmitCarrier`

### PR DESCRIPTION
**Pull request by maciejbaczmanski**
implemented `otPlatDiagRadioTransmitCarrier` functionality which lets user start and stop transmitting continuous carrier wave using `ot diag cw start` and `ot diag cw stop` commands.

implemented `otPlatRadioGetState` functionality to enable checking current state using `ot diag radio state` command.

disabled OT power calibration support as in Zephyr it is handled by Radio Driver, so it can't be handled on OT level.

### CODE REVIEW

File: modules/openthread/platform/radio.c [hunk 1]
```diff
@@ -631,7 +631,7 @@ otError otPlatRadioSleep(otInstance *aInstance)
 	    sState == OT_RADIO_STATE_RECEIVE ||
 	    sState == OT_RADIO_STATE_TRANSMIT) {
 		error = OT_ERROR_NONE;
-		if (radio_api->stop(radio_dev)) {
+		if (radio_api->stop(radio_dev) == 0) {
```
Comment (canisLupus1313) [inline modules/openthread/platform/radio.c:634]: Please consider keeping consistency along all stop calls and either compare them all to 0 or remove this comparison.

File: modules/openthread/platform/radio.c [hunk 2]
```diff
@@ -674,6 +674,36 @@ otError otPlatRadioReceiveAt(otInstance *aInstance, uint8_t aChannel,
...
+	if ((aEnable) && (sState == OT_RADIO_STATE_RECEIVE)) {
+		radio_api->set_txpower(radio_dev, get_transmit_power_for_channel(channel));
+
+		if (radio_api->continuous_carrier(radio_dev) != 0) {
+			return OT_ERROR_FAILED;
+		}
+
+		sState = OT_RADIO_STATE_TRANSMIT;
+	} else if ((!aEnable) && (sState == OT_RADIO_STATE_TRANSMIT)) {
+		otPlatRadioReceive(aInstance, channel);
```
Comment (canisLupus1313) [inline modules/openthread/platform/radio.c:692]: Why are You starting reception in transmit carrier function? And not returning any error?

Comment (maciejbaczmanski) [reply modules/openthread/platform/radio.c]: added returning of result `otPlatRadioReceive()`

File: modules/openthread/platform/radio.c (hunk 1 above)
Comment (maciejbaczmanski) [reply modules/openthread/platform/radio.c]: removed comparison to keep consistency with other api calls

**Review by canisLupus1313 — DISMISSED** [review]

File: include/zephyr/net/ieee802154_radio.h [hunk 3]
```diff
@@ -340,6 +340,9 @@ struct ieee802154_radio_api {
 	/** Stop the device */
 	int (*stop)(const struct device *dev);
 
+	/** Start continuous carrier wave transmission */
+	int (*continuous_carrier)(const struct device *dev);
```
Comment (rlubos) [inline include/zephyr/net/ieee802154_radio.h:344]: How do you leave this mode?

Plus this should be a separate commit, it's not directly related to OT.

File: drivers/ieee802154/ieee802154_nrf5.c [hunk 4]
```diff
@@ -722,6 +722,23 @@ static int nrf5_stop(const struct device *dev)
 	return 0;
 }
 
+#if NRF_802154_CARRIER_FUNCTIONS_ENABLED
```
Comment (rlubos) [inline drivers/ieee802154/ieee802154_nrf5.c:725]: I think this should get a Kconfig entry in `modules/hal_nordic/Kconfig`.

Plus, a change in a specific radio driver, also merits a separate commit.

File: modules/openthread/platform/platform-zephyr.h [hunk 5]
```diff
@@ -70,6 +70,11 @@ void platformUartPanic(void);
  */
 uint16_t platformRadioChannelGet(otInstance *aInstance);
 
+/**
+ * Start/ stop continuous carrier wave transmission.
```
Comment (rlubos) [inline modules/openthread/platform/platform-zephyr.h:74]:
```suggestion
 * Start/stop continuous carrier wave transmission.
```

File: modules/openthread/platform/radio.c [hunk 6]
```diff
@@ -631,9 +631,8 @@ otError otPlatRadioSleep(otInstance *aInstance)
 	    sState == OT_RADIO_STATE_RECEIVE ||
 	    sState == OT_RADIO_STATE_TRANSMIT) {
 		error = OT_ERROR_NONE;
-		if (radio_api->stop(radio_dev)) {
-			sState = OT_RADIO_STATE_SLEEP;
-		}
+		radio_api->stop(radio_dev);
+		sState = OT_RADIO_STATE_SLEEP;
```
Comment (rlubos) [inline modules/openthread/platform/radio.c:635]: Why this change? It's not listed in the commit message.

File: drivers/ieee802154/ieee802154_nrf5.c (hunk 4 above)
Comment (rlubos) [reply drivers/ieee802154/ieee802154_nrf5.c:725]: Indeed we should. Too bad I wasn't CC'ed on that.

File: modules/openthread/platform/radio.c (hunk 6 above)
Comment (maciejbaczmanski) [reply modules/openthread/platform/radio.c:635]: Forgot to add it to the message. This fixes wrong check of `radio_api->stop()` return value
in `otPlatRadioSleep()` (api returns 0 on success, so sState was never set to `OT_RADIO_STATE_SLEEP`). 
Updated commit message.

File: drivers/ieee802154/ieee802154_nrf5.c (hunk 4 above)
Comment (maciejbaczmanski) [reply drivers/ieee802154/ieee802154_nrf5.c:725]: Moved change to separate commit

File: include/zephyr/net/ieee802154_radio.h (hunk 3 above)
Comment (maciejbaczmanski) [reply include/zephyr/net/ieee802154_radio.h:344]: To leave this mode redio's state should be changed back to receive (as it is done by calling `otPlatRadioReceive()` in `platformRadioTransmitCarrier()`, which then calls `radio_api->start()`). Alternatively, changing radio's state to sleep (e.g. by `ot diag radio sleep` command) will also stop the transmission.

Moved change to a separate commit.

Comment (rlubos) [reply include/zephyr/net/ieee802154_radio.h:344]: Yeah, but from the 15.4 radio driver perspective, I think it's not clear. IMO it should be documented, that in order to leave this mode one should call `start()` to bring it back into recv state, or `stop()`.

Comment (maciejbaczmanski) [reply include/zephyr/net/ieee802154_radio.h:347]: Right. Added more info to comment above.

**Review by rlubos — COMMENTED** [review]
LGTM, but we should wait for https://github.com/zephyrproject-rtos/zephyr/pull/52600 and switch to use Kconfig options in the code instead.

**Review by canisLupus1313 — DISMISSED** [review]

File: drivers/ieee802154/ieee802154_nrf5.c [hunk 7]
```diff
@@ -724,6 +724,23 @@ static int nrf5_stop(const struct device *dev)
...
```
Comment (rlubos) [inline drivers/ieee802154/ieee802154_nrf5.c:727]: Please use `#if defined(CONFIG_NRF_802154_CARRIER_FUNCTIONS)`

File: drivers/ieee802154/ieee802154_nrf5.c [hunk 8]
```diff
@@ -1169,6 +1186,9 @@ static struct ieee802154_radio_api nrf5_radio_api = {
 	.set_txpower = nrf5_set_txpower,
 	.start = nrf5_start,
 	.stop = nrf5_stop,
+#if NRF_802154_CARRIER_FUNCTIONS_ENABLED
```
Comment (rlubos) [inline drivers/ieee802154/ieee802154_nrf5.c:1189]: ditto

**Review by rlubos — APPROVED** [review]

File: drivers/ieee802154/ieee802154_nrf5.c [hunk 9]
```diff
@@ -724,6 +724,23 @@ static int nrf5_stop(const struct device *dev)
 	return 0;
 }
 
+#if defined(CONFIG_NRF_802154_CARRIER_FUNCTIONS)
+static int nrf5_continuous_carrier(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	if (!nrf_802154_continuous_carrier()) {
```
Comment (edmont) [inline drivers/ieee802154/ieee802154_nrf5.c:732]: nit: file code style suggests bracket in new line.

File: drivers/ieee802154/ieee802154_nrf5.c [hunk 10]
```diff
@@ -724,6 +724,23 @@ static int nrf5_stop(const struct device *dev)
...
+		LOG_ERR("Failed to enter continuous carrier state");
+		return -EIO;
+	}
+
+	LOG_DBG("Continuous carrier wave transmission started (channel: %d)",
+		nrf_802154_channel_get());
```
Comment (edmont) [inline drivers/ieee802154/ieee802154_nrf5.c:738]: nit: move to line above.

**Review by edmont — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
